### PR TITLE
Support grpclb address in the c-ares resolver 

### PR DESCRIPTION
**Pull request by y-zeng**
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/11237)

### CODE REVIEW

**Comment by markdroth** [comment]
Overall, this looks great!  Just a few structural suggestions.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 66 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWC82DbQRk77QhYBno:-KkWC82DbQRk77QhYBnp:b-9wk6j7) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L66)):*
>   union {
>     grpc_resolved_addresses **addrs;
>     grpc_lb_addresses **lb_addrs;
>   } addrs_out;
>   /** if true, the output addresses are in the format of grpc_lb_addresses,
>       otherwise they are in the format of grpc_resolved_addresses */
>   bool lb_addrs_out;

Instead of plumbing this difference through the whole code, I suggest having the resolver code always generate a `grpc_lb_addresses` struct, and then convert from that to `grpc_resolved_addresses` inside of `grpc_resolve_address_ares_impl()`.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 183 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWJBWJLhg5hSMfsrbW:-KkWJBWJLhg5hSMfsrbX:b5e076v) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L183)):*
> ```C
>       for (i = prev_naddr; i < (*lb_addresses)->num_addresses; i++) {
>         memset(&(*lb_addresses)->addresses[i], 0, sizeof(grpc_lb_address));
>         if (hostent->h_addrtype == AF_INET6) {
> ```

I suggest making this a switch statement, so that the two sections (IPv6 and IPv4) are self-documenting.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 315 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWGpDwXEe7eFocWyGN:-KkWGpDxS8FEqN3QEP49:b-4scdg5) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L315)):*
> ```C
>       }
>     }
> ```

Nit: Please remove blank lines within functions.  (Both here and throughout.)

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 432 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWFrCwnqIOfvARNM0M:-KkWFrCxHYWH7Q730wDA:b-6vqst5) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L432)):*
> ```C
>   resolve_address_ares_impl(exec_ctx, name, default_port, interested_parties,
>                             on_done, (void **)addrs,
>                             true /* is_lb_addrs_out */);
> ```

Instead of hard-coding this here, how about adding a `bool check_grpclb` parameter to `grpc_dns_lookup_ares()` (which you've called `grpc_resolve_grpclb_address_ares()`)?  That way, we can just have `grpc_resolve_address_ares_impl()` call `grpc_dns_lookup_ares()` and won't need an extra level of indirection.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h, line 62 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWDT25UorB-yG1bud2:-KkWDT26R00nVeZS2SGI:b-rdcp1o) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h#L62)):*
> ```objc
>   scheduled with \a exec_ctx, it must not try to acquire locks that are being
>   held by the caller. */
> void grpc_resolve_grpclb_address_ares(grpc_exec_ctx *exec_ctx, const char *addr,
> ```

Once you merge in the changes from #10706, which I merged this morning, you can just change the `grpc_dns_lookup_ares()` function that I wrote to return a `grpc_lb_addresses` struct instead of a `grpc_resolved_addresses` struct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/11237#-:-KkWKgv8utV688qtioSt:b-ixlw8j)*

**Comment by y-zeng** [comment]
Thanks a lot for your review!

---

Review status: 1 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 66 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWC82DbQRk77QhYBno:-KkXxJbCwSZpbIZabPOZ:b-glwydd) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L66)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

>   union {
>     grpc_resolved_addresses **addrs;
>     grpc_lb_addresses **lb_addrs;
>   } addrs_out;
>   /** if true, the output addresses are in the format of grpc_lb_addresses,
>       otherwise they are in the format of grpc_resolved_addresses */
>   bool lb_addrs_out;

Instead of plumbing this difference through the whole code, I suggest having the resolver code always generate a `grpc_lb_addresses` struct, and then convert from that to `grpc_resolved_addresses` inside of `grpc_resolve_address_ares_impl()`.
</blockquote></details>

As `grpc_resolve_address_ares_impl()` and `grpc_dns_lookup_ares()` are async operations, we'll need an extra level of callback for `grpc_resolve_address_ares_impl()` to convert the result from `grpc_lb_addresses` to `grpc_resolved_addresses`. It will also bring some extra memory allocs, frees, and copies.

It might be good to save this extra level of callback and the memory operations, by still having these fields in `grpc_ares_request` and handling the conversion in `on_hostbyname_done_cb`.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 183 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWJBWJLhg5hSMfsrbW:-KkXp9N1PGesxTwbLs5i:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L183)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I suggest making this a switch statement, so that the two sections (IPv6 and IPv4) are self-documenting.
</blockquote></details>

Done.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 315 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWGpDwXEe7eFocWyGN:-KkXnNoHHBH5Kxocosia:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L315)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Nit: Please remove blank lines within functions.  (Both here and throughout.)
</blockquote></details>

Done.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 432 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWFrCwnqIOfvARNM0M:-KkXnaLqzaodOKd8n94D:bs7wumx) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L432)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Instead of hard-coding this here, how about adding a `bool check_grpclb` parameter to `grpc_dns_lookup_ares()` (which you've called `grpc_resolve_grpclb_address_ares()`)?  That way, we can just have `grpc_resolve_address_ares_impl()` call `grpc_dns_lookup_ares()` and won't need an extra level of indirection.
</blockquote></details>

Done. 

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h, line 62 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWDT25UorB-yG1bud2:-KkY5ZFacU1y5fiFX5YG:b64sae4) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h#L62)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Once you merge in the changes from #10706, which I merged this morning, you can just change the `grpc_dns_lookup_ares()` function that I wrote to return a `grpc_lb_addresses` struct instead of a `grpc_resolved_addresses` struct.
</blockquote></details>

Changed `grpc_dns_lookup_ares()` to handle both `grpc_lb_addresses` and `grpc_resolved_addresses`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/11237#-:-KkXmohPh_Or8YIMP3OO:b-i80v59)*

**Comment by markdroth** [comment]
Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 66 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWC82DbQRk77QhYBno:-KkkqG_xHk1F1PAAFnbu:ba566h0) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L66)):*
<details><summary><i>Previously, y-zeng (Yuchen Zeng) wrote…</i></summary><blockquote>

As `grpc_resolve_address_ares_impl()` and `grpc_dns_lookup_ares()` are async operations, we'll need an extra level of callback for `grpc_resolve_address_ares_impl()` to convert the result from `grpc_lb_addresses` to `grpc_resolved_addresses`. It will also bring some extra memory allocs, frees, and copies.

It might be good to save this extra level of callback and the memory operations, by still having these fields in `grpc_ares_request` and handling the conversion in `on_hostbyname_done_cb`.
</blockquote></details>

I'll leave it up to you, but I suspect that the extra level of indirection would be the better approach here.  It's true that the extra level of indirection would be slightly less efficient, but it's not in the fast path, so it shouldn't really hurt anything.  And I think the benefit of the cleaner code would be worth the trade-off.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/11237)*

**Comment by y-zeng** [comment]
Review status: 2 of 5 files reviewed at latest revision, all discussions resolved.

---

*[src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c, line 66 at r1](https://reviewable.io:443/reviews/grpc/grpc/11237#-KkWC82DbQRk77QhYBno:-KkwpS6PHdUik2CRRLOu:bysel2h) ([raw file](https://github.com/grpc/grpc/blob/72f70a5a5d3a438a00e3ac73a908344693c274e9/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c#L66)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I'll leave it up to you, but I suspect that the extra level of indirection would be the better approach here.  It's true that the extra level of indirection would be slightly less efficient, but it's not in the fast path, so it shouldn't really hurt anything.  And I think the benefit of the cleaner code would be worth the trade-off.
</blockquote></details>

I've thought about the options, and moved the conversion code to the callback of ` grpc_resolve_address_ares_impl()` as you suggested.  After c-ares resolver becomes the default one, client channel should only use `grpc_ares_dns_lookup()`. The extra level of indirection will only affect the credential verification process. Thanks a lot for your suggestion!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/11237)*

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c [hunk 1]
```diff
@@ -110,75 +121,104 @@ static void grpc_ares_request_unref(grpc_exec_ctx *exec_ctx,
...
-      (*addresses)->naddrs = 0;
-      (*addresses)->addrs = NULL;
+    grpc_lb_addresses **lb_addresses = r->lb_addrs_out;
+    if (*lb_addresses == NULL) {
+      *lb_addresses = grpc_lb_addresses_create(0, NULL);
     }
-    size_t prev_naddr = (*addresses)->naddrs;
+    size_t prev_naddr = (*lb_addresses)->num_addresses;
     size_t i;
     for (i = 0; hostent->h_addr_list[i] != NULL; i++) {
```
Comment (apolcyn) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c]: nit: no-op in this loop?

Comment (apolcyn) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.c]: sorry nevermind, scratched comment

**Comment by y-zeng** [comment]
Known issues:
Asan C++: #11120
Basic Tests Linux: #11122
Basic Tests Mac: #11223, #9380

**Comment by y-zeng** [comment]
Filed #11424 for https://grpc-testing.appspot.com/job/gRPC_pull_requests_tsan_c/5092/testReport/junit/(root)/c_linux_tsan_native/bins_tsan_sequential_connectivity_test_GRPC_POLL_STRATEGY_epollsig/

**Comment by markdroth** [comment]
Still LGTM.  Thanks for doing this!

**Comment by y-zeng** [comment]
Thanks a lot for your review!

**Pull request merged**